internal/util/secretskms.go

```diff
@@ -91,24 +100,75 @@ var _ = RegisterKMSProvider(KMSProvider{
 	Initializer: initSecretsMetadataKMS,
 })
 
-// initSecretsMetadataKMS initializes a SecretsMetadataKMS that wraps a
-// SecretsKMS, so that the passphrase from the StorageClass secrets can be used
+// initSecretsMetadataKMS initializes a SecretsMetadataKMS that wraps a SecretsKMS,
+// so that the passphrase from the user provided or StorageClass secrets can be used
 // for encrypting/decrypting DEKs that are stored in a detached DEKStore.
 func initSecretsMetadataKMS(args KMSInitializerArgs) (EncryptionKMS, error) {
-	eKMS, err := initSecretsKMS(args.Secrets)
+	var (
+		smKMS                SecretsMetadataKMS
+		encryptionPassphrase string
+		ok                   bool
+		err                  error
+	)
+
+	encryptionPassphrase, err = smKMS.fetchEncryptionPassphrase(
+		args.Config, args.Tenant)
 	if err != nil {
-		return nil, err
+		if !errors.Is(err, errConfigOptionMissing) {
+			return nil, err
+		}
+		// if 'userSecret' option is not specified, fetch encryptionPassphrase
+		// from storageclass secrets.
+		encryptionPassphrase, ok = args.Secrets[encryptionPassphraseKey]
```

Comment:
Is there a reason to take this order when checking for the encryption passphrase? In case it needs to be fetched from the StorageClass Secret, this now adds a call to Kubernetes that fails if the tenant does not have the secret (and even panics on non-Kubernetes).

Reply:
It will call k8s api for secrets only if `userSecret` key is set in kms config.

If `userSecret` key is not found, it behaves in the same old with finding `encryptionPassphrase` in storage class secrets.